### PR TITLE
[Concurrency] SIL: add hop_to_executor instruction

### PR DESCRIPTION
**Pull request by eeckstein**
This instructions ensures that all instructions, which need to run on the specified executor actually run on that executor.
For details see the description in SIL.rst.

This instruction also creates an "SSA edge". This means it returns its operand and all subsequent uses of the operand actor should use the instruction's result.
I'm not 100% sure if this is the right design decision, because it complicates the SIL a bit (e.g. it can introduce phi-arguments in case of control flow).
On the other hand it feels right. For example it automatically enforces the dependency between this instruction and subsequent actor uses.

rdar://problem/70298801

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test linux

**Comment by rjmccall** [comment]
I don't know about the SSA thing.  Is that concretely necessary for something?  Can we not validate that otherwise?

**Comment by atrick** [comment]
Generally, any side effects will be ordered with respect to `hop-to-executor`. The only situation I can think of for an SSA edge:
- reading something from the Actor
- a requirement that the read to occur after the hop (I can't imagine why)
- SILGen ensures that such reads are "protected" by a hop instruction
- SIL analysis can't be "aware" that the hop's operand and result are the same Actor (seems unfortunate)

**Comment by gottesmm** [comment]
@atrick is it safe to assume that SILGen will always protect such a read by a "hop" instruction. If we create an invariant in the IR we can guarantee that and define away mistakes.

**Comment by atrick** [comment]
We could statically enforce that SIL can't read from an actor unless it's running on the Actor's executor, but I don't know why that would be desirable.

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
I thought about it: I think it's simpler to make hop_to_executor a non-value instruction. I pushed a new version

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Review by DougGregor — APPROVED** [review]

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test

File: docs/SIL.rst [hunk 1]
```diff
@@ -2885,6 +2885,27 @@ initialized with the resume value, and that value is then owned by the current
...
+
+  sil-instruction ::= 'hop_to_executor' sil-operand
+
+  hop_to_executor %0 : $T
+
+  // $T must conform to the Actor protocol
+
+Ensures that all instructions, which need to run on the actor's executor
+actually run on that executor.
+This instruction can only be used inside an ``@async`` function.
```
Comment (zoecarver) [inline docs/SIL.rst:2901]: Could there be a SILVerifier check for this?

File: docs/SIL.rst [hunk 2]
```diff
@@ -2885,6 +2885,27 @@ initialized with the resume value, and that value is then owned by the current
...
+
+Checks if the current executor is the one which is bound to the operand actor.
+If not, begins a suspension point and enqueues the continuation to the executor
+which is bound to the operand actor.
+
+The operand is a guaranteed operand, i.e. no consumed.
```
Comment (zoecarver) [inline docs/SIL.rst:2907]: "not consumed"

File: docs/SIL.rst (hunk 1 above)
Comment (zoecarver) [reply docs/SIL.rst:2901]: Maybe we could also check the operand is an actor?

Comment (eeckstein) [reply docs/SIL.rst:2901]: Yes, there should definitely be verifier checks for this instruction. It's an open to-do.

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test macOS

**Comment by eeckstein** [comment]
@swift-ci smoke test macOS

**Pull request merged**